### DISCUSSION TITLE
issue: how to properly type a controlled element with the Validate generic in 7.42 how requiring the formValues type

### DISCUSSION THREAD

**Discussion by doberkofler** [react-hook-form/react-hook-form#9769] [Q&A] [answered]
### Version Number

7.42.0

### Codesandbox/Expo snack

Not possible as there is no 7.42.0 dependency yet

### Steps to reproduce

When type checking the following example a `Generic type 'Validate' requires 2 type argument(s).` error is reported in 7.42.0 in the line 26 `validate?: Validate<string>,`.

Unfortunately I see no way to keep the `ControlledTextField` component independent but still correctly typed after the changes introduced in 7.42.0

```typescript
import React from 'react';
import {useController} from 'react-hook-form';
import TextField from '@mui/material/TextField';
//import {elementDefaults} from './defaults';

import type {Control, FieldValues, FieldPath, ValidationRule, Validate} from 'react-hook-form';
import type {TextFieldProps} from '@mui/material/TextField';

// inline for test
type PropsBaseType<TFieldValues extends FieldValues> = {
	control: Control<TFieldValues>,
	name: FieldPath<TFieldValues>,
	disabled?: boolean,
	placeholder?: string,
	label: string,
	helperText?: string,
	errorText?: string,
	autoFocus?: boolean,
	required?: boolean,
};

type PropsTextType<TFieldValues extends FieldValues> = PropsBaseType<TFieldValues> & {
	minLength?: ValidationRule<number>,
	maxLength?: ValidationRule<number>,
	pattern?: ValidationRule<RegExp>,
	validate?: Validate<string>,
};

export const ControlledTextField = <TFieldValues extends FieldValues>({
	control,
	name,
	disabled = false,
	label = '',
	placeholder = '',
	helperText = '',
	autoFocus = false,
	required = false,
	minLength,
	maxLength,
	pattern,
	validate,
}: PropsTextType<TFieldValues>): JSX.Element => {
	const {field, fieldState} = useController({
		control,
		name,
		rules: {
			required,
			minLength,
			maxLength,
			pattern,
			validate,
		},
	});
	const [value, setValue] = React.useState<string>(field.value);
	const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		field.onChange(e.target.value);
		setValue(e.target.value);
	};
	const fieldProps: TextFieldProps = {
		disabled,
		label,
		placeholder,
		error: Boolean(fieldState.error),
		helperText: fieldState.error ? 'Error' : helperText,
		autoFocus,
		value,
		name: field.name,
		onChange: handleChange,
		onBlur: field.onBlur,
		inputRef: field.ref,
		//...elementDefaults,
	};

	return <TextField {...fieldProps}/>;
};
```

### Expected behaviour

Basically it should remain possible to specify a Validate rule for an individual component without having to know about the form type, but I guess I just miss some Typescript trickery ;-)

### Relevant log output

```shell
eslint:

???.tsx
  51:4  warning  Unsafe assignment of an `any` value  @typescript-eslint/no-unsafe-assignment


tsc:
```bash
???.tsx:26:13 - error TS2314: Generic type 'Validate' requires 2 type argument(s).

26  validate?: Validate<string>,
               ~~~~~~~~~~~~~~~~
```
```


### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by doberkofler** [comment, marked as answer]
Sorry. I just completely missed it after having written this code quite some time ago. It's so simple:

```typescript
type PropsTextType<TFieldValues extends FieldValues> = PropsBaseType<TFieldValues> & {
	minLength?: ValidationRule<number>,
	maxLength?: ValidationRule<number>,
	pattern?: ValidationRule<RegExp>,
	validate?: Validate<string, TFieldValues>,
};
```

Solved!
